feat: optimize truncate(col) == value to startsWith predicate

shangxinli opened the pull request:
Rewrite truncate(col) == "value" predicates to col STARTS_WITH "value" for string columns. This enables better predicate pushdown to storage formats and efficient use of prefix indexes.

The optimization only applies when:
- Operation is equality
- Term is a truncate transform
- Literal is a string type

Added tests to verify correct application and edge cases.

wgtmac commented:
There are some merge conflicts...

shangxinli commented:
Fixed

wgtmac commented on src/iceberg/expression/predicate.cc:
```diff
@@ -143,6 +145,27 @@ bool IsFloatingType(TypeId type) {
+///
+/// \param str The UTF-8 encoded string
+/// \return The number of code points (not bytes, not graphemes)
+inline int32_t CountUTF8CodePoints(std::string_view str) {
```
Perhaps this should be moved to `string_util.h` or `truncate_util.h`

wgtmac commented on src/iceberg/expression/predicate.cc:
```diff
@@ -246,7 +269,69 @@ Result<std::shared_ptr<Expression>> UnboundPredicate<B>::BindLiteralOperation(
+        !literal.IsNull()) {  // Null safety: skip null literals
+
+      // Extract width parameter using type-safe API
+      auto width_opt = transform_term->transform()->param();
```
I think we can greatly simplify this line and below.

We can add `transform_func` like below.

```cpp
class ICEBERG_EXPORT BoundTransform : public BoundTerm {
 public:
  const std::shared_ptr<TransformFunction>& transform_func() const { return transform_func_; }
};
```

Then we just need to call `transform_term->transform_func()->Transform(literal)` and check if its return value is same as `literal`. We don't even need `CountUTF8CodePoints` above.

shangxinli committed:
feat: optimize truncate(col) == value to startsWith predicate

Implements optimization to rewrite truncate equality predicates as
startsWith for better predicate pushdown and index usage.

The optimization applies when:
- Operation is equality
- Term is a truncate transform on string column
- Literal has exactly the truncate width in UTF-8 code points

Implementation uses transform_func()->Transform() to validate that:
1. truncate(literal) == literal (literal is compatible)
2. truncate(literal + 'x') == literal (literal has exact width)

This approach leverages the transform function without duplicating
UTF-8 code point counting logic.

zhjwpku commented on src/iceberg/expression/predicate.cc:
```diff
@@ -286,7 +289,52 @@ Result<std::shared_ptr<Expression>> UnboundPredicate<B>::BindLiteralOperation(
+
+      // Apply truncate transform to the literal and check if result matches
+      // This verifies the literal is compatible with the truncate operation
+      auto transformed_result = transform_term->transform_func()->Transform(literal);
```
I'm wondering whether we can simply count the code points in the literal and compare that with `TruncateTransform.width()`. If they match, we can rewrite it as a `StartsWith`. We don't need to transform the extended_literal then.

zhjwpku replied:
Also, if the number of code points in the literal is greater than `TruncateTransform.width()`, we can rewrite it as `AlwaysFalse`. I created PR [1] for the Java implementation to do this.

[1] https://github.com/apache/iceberg/pull/16383
